devito/function.py

```diff
@@ -330,50 +284,67 @@ def argument_values(self, alias=None, **kwargs):
                 # Set new values and re-derive defaults
                 values[key.name] = new._data_buffer
-                values.update(new.argument_defaults(alias=key).reduce_all())
+                values.update(new._arg_defaults(alias=key).reduce_all())
             else:
                 # We've been provided a pure-data replacement (array)
                 values[key.name] = new
                 # Add value overrides for all associated dimensions
                 for i, s, o in zip(self.indices, new.shape, self.staggered):
-                    values.update(i.argument_defaults(size=s+o))
+                    values.update(i._arg_defaults(size=s+o-sum(self._offset_domain[i])))
 
         return values
 
+    def _arg_check(self, args, intervals):
+        """
+        :raises InvalidArgument: If, given the runtime arguments ``args``, an
+                                 out-of-bounds access will be performed.
+        """
+        if self.name not in args:
+            raise InvalidArgument("No runtime value for `%s`" % self.name)
+        key = args[self.name]
+        if len(key.shape) != self.ndim:
+            raise InvalidArgument("Shape %s of runtime value `%s` does not match "
+                                  "dimensions %s" % (key.shape, self.name, self.indices))
+        if key.dtype != self.dtype:
+            warning("Data type %s of runtime value `%s` does not match the "
+                    "Function data type %s" % (key.dtype, self.name, self.dtype))
+        for i, s in zip(self.indices, key.shape):
+            i._arg_check(args, s, intervals[i])
+
 
 class Function(TensorFunction):
-    """Data object for spatially varying data acting as a :class:`SymbolicFunction`.
+    """A :class:`TensorFunction` providing operations to express
+    finite-difference approximation. A ``Function`` encapsulates
+    space-varying data; for time-varying data, use :class:`TimeFunction`.
 
     :param name: Name of the symbol
     :param grid: :class:`Grid` object from which to infer the data shape
                  and :class:`Dimension` indices.
+    :param space_order: Discretisation order for space derivatives. By default,
+                        ``space_order`` points are available on both sides of
+                        a generic point of interest, including those on the grid
+                        border. Sometimes, fewer points may be necessary; in
+                        other cases, depending on the PDE being approximated,
+                        more points may be necessary. In such cases, one
+                        can pass a 3-tuple ``(o, lp, rp)`` instead of a single
+                        integer representing the discretization order. Here,
+                        ``o`` is the discretization order, while ``lp`` and ``rp``
+                        indicate how many points are expected on left (``lp``)
+                        and right (``rp``) of a point of interest.
```

Comment:
- I'm not sure I like that as part of `space_order'.
- `space_order` as a tuple would makes sense for a different order in each of the dimensions (e.g 2nd in x and 4th in z).

Reply:
agreed, but this is exactly what we have in ``master``. Let's change it another day (volunteers welcome)